icarusalg/Utilities/TrackTimeInterval.h

```diff
@@ -0,0 +1,538 @@
+   * @return time range including all `recob::Hit` in the collection
+   * @see `timeRange(BIter, EIter) const`
+   * 
+   * See `timeRange(BIter begin, EIter end) const` for details.
+   * 
+   * Type requirements
+   * ------------------
+   * 
+   * `HitColl` must be a forward-iterable sequence whose elements are each one
+   * compatible with a `timeRange()` call.
+   * 
+   */
+  template <typename HitColl>
+  TimeRange timeRangeOfHits(HitColl const& hits) const;
+  
+  
+    private:
+  friend class TrackTimeIntervalMaker;
+  
+  using centimeters = ::util::quantities::intervals::centimeters;
+  struct TimeLimits_t {
+    centimeters driftDistance;
+  };
+  
+  struct GeometryCache_t {
+    using LimitsCache_t = geo::PlaneDataContainer<TimeLimits_t>;
+    using TPCtoSetMap_t = geo::TPCDataContainer<readout::TPCsetID>;
+    
+    /// Cached times for all the planes in the detector.
+    LimitsCache_t limits;
+    
+    /// Cached dimensions for a TPC set data container.
+    std::array<unsigned int, 2U> TPCsetDims;
+    
+    /// TPC to TPC set map.
+    TPCtoSetMap_t TPCtoSet;
+    
+    /// Returns the cached number of cryostats in the detector.
+    unsigned int Ncryostats() const { return TPCsetDims[0]; }
+    
+  }; // GeometryCache_t
+  
+  
+  /// Constructor: uses a provided cache instead of creating it.
+  TrackTimeInterval(
+    GeometryCache_t geomCache,
+    detinfo::DetectorPropertiesData detProp,
+    detinfo::DetectorTimings detTimings
+    );
+  
+  
+  /// Local copy of detector properties.
+  detinfo::DetectorPropertiesData const fDetProp;
+  
+  /// Local copy of the timing conversion utility (including `DetectorClocks`).
+  detinfo::DetectorTimings const fDetTimings;
+  
+  double const fDriftVelocity; ///< Detector drift velocity [cm/us]
+  
+  GeometryCache_t const fGeomCache; ///< Cached geometry information.
+  
+  
+  /// Creates a geometry cache.
+  static GeometryCache_t buildGeomCache(geo::GeometryCore const& geom);
+  
+  /// Returns the time limits of all the planes in the specified TPC.
+  static GeometryCache_t::LimitsCache_t extractTimeLimits
+    (geo::GeometryCore const& geom);
+  
+  static GeometryCache_t::TPCtoSetMap_t extractTPCtoSetMap
+    (geo::GeometryCore const& geom);
+  
+  
+  /**
+   * @brief Merges the ranges from all the TPC sets in the detector.
+   * @param TPCsetRanges container of allowed range from each TPC set
+   * @return a single allowed time range
+   * 
+   * As an input, each TPC set contributes its hypothesis of allowed time range;
+   * that hypothesis is invalid (`!isValid()`) if there was no information in
+   * the TPC set.
+   * The combination assumes that all the hits belong to the same activity and
+   * that the time of all that activity is only one.
+   * Contributions from different cryostats are intersected, while within each
+   * cryostat the combination is made in a detector-specific way (see
+   * `mergeTPCsetRanges_SBN()`.
+   */
```

Comment:
Doxygen comments here and in L315-325 are mixed up and possibly also from a previous version?

Reply:
Mixed up, not sure if from a previous version but I do not exclude. Thank you for the catch!
Fixed in a new commit.